{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import  precision_score, recall_score, f1_score, roc_curve, auc, precision_recall_curve\n",
    "from difflib import SequenceMatcher"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.read_csv('./db.csv')\n",
    "df=df.iloc[:111]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df.drop(['Entry','Reviewed', 'Entry Name', 'Gene Names', 'Gene Ontology (biological process)', 'Gene Ontology (molecular function)'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Truncated endolysin</td>\n",
       "      <td>Staphylococcus phage 812</td>\n",
       "      <td>284</td>\n",
       "      <td>MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Putative holin</td>\n",
       "      <td>Staphylococcus phage S24-1</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Putative holin</td>\n",
       "      <td>Staphylococcus phage S13'</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Putative holin</td>\n",
       "      <td>Staphylococcus phage S25-3</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Putative holin</td>\n",
       "      <td>Staphylococcus phage S25-4</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Bifunctional endolysin</td>\n",
       "      <td>Staphylococcus phage SpT152</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Bifunctional endolysin</td>\n",
       "      <td>Staphylococcus phage vB_SpsM_WIS42</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Bi-functional endolysin</td>\n",
       "      <td>Staphylococcus phage Sebago</td>\n",
       "      <td>624</td>\n",
       "      <td>MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Bacteriophage phi 11 holin homologue (ORF3)</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>57</td>\n",
       "      <td>MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Bacteriophage phi 11 holin homologue (ORF3)</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>49</td>\n",
       "      <td>MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Protein names  \\\n",
       "0                            Truncated endolysin   \n",
       "1                                 Putative holin   \n",
       "2                                 Putative holin   \n",
       "3                                 Putative holin   \n",
       "4                                 Putative holin   \n",
       "..                                           ...   \n",
       "106                       Bifunctional endolysin   \n",
       "107                       Bifunctional endolysin   \n",
       "108                      Bi-functional endolysin   \n",
       "109  Bacteriophage phi 11 holin homologue (ORF3)   \n",
       "110  Bacteriophage phi 11 holin homologue (ORF3)   \n",
       "\n",
       "                                              Organism  Length  \\\n",
       "0                             Staphylococcus phage 812     284   \n",
       "1                           Staphylococcus phage S24-1     140   \n",
       "2                            Staphylococcus phage S13'     140   \n",
       "3                           Staphylococcus phage S25-3     167   \n",
       "4                           Staphylococcus phage S25-4     167   \n",
       "..                                                 ...     ...   \n",
       "106                        Staphylococcus phage SpT152     629   \n",
       "107                 Staphylococcus phage vB_SpsM_WIS42     629   \n",
       "108                        Staphylococcus phage Sebago     624   \n",
       "109  Staphylococcus phage phi11 (Bacteriophage phi-11)      57   \n",
       "110  Staphylococcus phage phi11 (Bacteriophage phi-11)      49   \n",
       "\n",
       "                                              Sequence  \n",
       "0    MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...  \n",
       "1    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...  \n",
       "2    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...  \n",
       "3    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...  \n",
       "4    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...  \n",
       "..                                                 ...  \n",
       "106  MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...  \n",
       "107  MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...  \n",
       "108  MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...  \n",
       "109  MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...  \n",
       "110  MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH  \n",
       "\n",
       "[111 rows x 4 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "for name in df['Protein names']:\n",
    "    df['Protein names'] = np.where(df['Protein names'].str.contains('endolysin'), 'endolysin', df['Protein names'])\n",
    "    df['Protein names'] = np.where(df['Protein names'].str.contains('Endolysin'), 'endolysin', df['Protein names'])\n",
    "    df['Protein names'] = np.where(df['Protein names'].str.contains('Holin'), 'holin', df['Protein names'])\n",
    "    df['Protein names'] = np.where(df['Protein names'].str.contains('holin'), 'holin', df['Protein names'])\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins=df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage 812</td>\n",
       "      <td>284</td>\n",
       "      <td>MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S24-1</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S13'</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S25-3</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S25-4</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage SpT152</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage vB_SpsM_WIS42</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage Sebago</td>\n",
       "      <td>624</td>\n",
       "      <td>MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>57</td>\n",
       "      <td>MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>49</td>\n",
       "      <td>MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Protein names                                           Organism  Length  \\\n",
       "0       endolysin                           Staphylococcus phage 812     284   \n",
       "1           holin                         Staphylococcus phage S24-1     140   \n",
       "2           holin                          Staphylococcus phage S13'     140   \n",
       "3           holin                         Staphylococcus phage S25-3     167   \n",
       "4           holin                         Staphylococcus phage S25-4     167   \n",
       "..            ...                                                ...     ...   \n",
       "106     endolysin                        Staphylococcus phage SpT152     629   \n",
       "107     endolysin                 Staphylococcus phage vB_SpsM_WIS42     629   \n",
       "108     endolysin                        Staphylococcus phage Sebago     624   \n",
       "109         holin  Staphylococcus phage phi11 (Bacteriophage phi-11)      57   \n",
       "110         holin  Staphylococcus phage phi11 (Bacteriophage phi-11)      49   \n",
       "\n",
       "                                              Sequence  \n",
       "0    MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...  \n",
       "1    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...  \n",
       "2    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...  \n",
       "3    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...  \n",
       "4    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...  \n",
       "..                                                 ...  \n",
       "106  MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...  \n",
       "107  MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...  \n",
       "108  MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...  \n",
       "109  MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...  \n",
       "110  MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH  \n",
       "\n",
       "[111 rows x 4 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['endolysin', 'holin'], dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Protein names'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "sequences=df['Sequence']\n",
    "def calculate_ngram_frequencies(sequence, n):\n",
    "    ngram_freq = {}\n",
    "    for i in range(len(sequence) - n + 1):\n",
    "        ngram = sequence[i:i+n]\n",
    "        ngram_freq[ngram] = ngram_freq.get(ngram, 0) + 1\n",
    "    return ngram_freq\n",
    "\n",
    "# Defining the function to create feature vectors\n",
    "def create_feature_vector(row, n, feature_vector):\n",
    "    sequence = row['Sequence']\n",
    "    ngram_freq = calculate_ngram_frequencies(sequence, n)\n",
    "    # Initializing the feature vector with zeros\n",
    "    vector = feature_vector.copy()\n",
    "    # Updating the feature vector with the frequency of each n-gram\n",
    "    for ngram, freq in ngram_freq.items():\n",
    "        vector[ngram] = freq\n",
    "    return pd.Series(vector)\n",
    "\n",
    "data = {'Sequence': df['Sequence'].tolist()}  # Extracting the 'Sequence' column as a list\n",
    "\n",
    "# Converting the list of sequences to a dictionary with key 'Sequence'\n",
    "data_dict = {'Sequence': data['Sequence']}\n",
    "\n",
    "datF = pd.DataFrame(data_dict)\n",
    "# Defining the value of n for n-grams\n",
    "n = 3\n",
    "\n",
    "# Generating the feature vector template\n",
    "amino_acids = \"ACDEFGHIKLMNPQRSTVWY\"  # A string representing the 20 standard amino acids\n",
    "feature_vector = {\"\".join(aa): 0 for aa in itertools.product(amino_acids, repeat=n)}\n",
    "\n",
    "# Applying the function to create feature vectors for each row\n",
    "feature_vectors = datF.apply(lambda row: create_feature_vector(row, n, feature_vector), axis=1)\n",
    "\n",
    "# Combining sequence and feature vectors into a single DataFrame\n",
    "result_df = pd.concat([datF['Sequence'], feature_vectors], axis=1)\n",
    "df = pd.concat([df, result_df.drop(columns=['Sequence'])], axis=1, join='inner')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAD</th>\n",
       "      <th>AAE</th>\n",
       "      <th>AAF</th>\n",
       "      <th>AAG</th>\n",
       "      <th>...</th>\n",
       "      <th>YYM</th>\n",
       "      <th>YYN</th>\n",
       "      <th>YYP</th>\n",
       "      <th>YYQ</th>\n",
       "      <th>YYR</th>\n",
       "      <th>YYS</th>\n",
       "      <th>YYT</th>\n",
       "      <th>YYV</th>\n",
       "      <th>YYW</th>\n",
       "      <th>YYY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage 812</td>\n",
       "      <td>284</td>\n",
       "      <td>MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S24-1</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S13'</td>\n",
       "      <td>140</td>\n",
       "      <td>MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S25-3</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage S25-4</td>\n",
       "      <td>167</td>\n",
       "      <td>MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage SpT152</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage vB_SpsM_WIS42</td>\n",
       "      <td>629</td>\n",
       "      <td>MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>Staphylococcus phage Sebago</td>\n",
       "      <td>624</td>\n",
       "      <td>MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>57</td>\n",
       "      <td>MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>holin</td>\n",
       "      <td>Staphylococcus phage phi11 (Bacteriophage phi-11)</td>\n",
       "      <td>49</td>\n",
       "      <td>MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 8004 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Protein names                                           Organism  Length  \\\n",
       "0       endolysin                           Staphylococcus phage 812     284   \n",
       "1           holin                         Staphylococcus phage S24-1     140   \n",
       "2           holin                          Staphylococcus phage S13'     140   \n",
       "3           holin                         Staphylococcus phage S25-3     167   \n",
       "4           holin                         Staphylococcus phage S25-4     167   \n",
       "..            ...                                                ...     ...   \n",
       "106     endolysin                        Staphylococcus phage SpT152     629   \n",
       "107     endolysin                 Staphylococcus phage vB_SpsM_WIS42     629   \n",
       "108     endolysin                        Staphylococcus phage Sebago     624   \n",
       "109         holin  Staphylococcus phage phi11 (Bacteriophage phi-11)      57   \n",
       "110         holin  Staphylococcus phage phi11 (Bacteriophage phi-11)      49   \n",
       "\n",
       "                                              Sequence  AAA  AAC  AAD  AAE  \\\n",
       "0    MAKTQAEINKRLDAYAKGTVDSPYRVKKATSYDPSFGVMEAGAIDA...    0    0    0    0   \n",
       "1    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...    0    0    0    0   \n",
       "2    MNEVKLRFTDTEAFHMFIYAGDLKLLYFLFVLMIVDIVTGFAKAIK...    0    0    0    0   \n",
       "3    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...    0    0    0    0   \n",
       "4    MANETKQPKVVGGINLSTRTKSKTFWVAIISAVALFANQITGAFGL...    0    0    0    0   \n",
       "..                                                 ...  ...  ...  ...  ...   \n",
       "106  MALPKTGKPTAKQVADWAISLIGSGVDVDGYYGRQCWDLSNYIFNR...    0    0    0    0   \n",
       "107  MALPKTGKPTAKQVVDWAINLIGSGVDVDGYYGRQCWDLPNYIFNR...    0    0    0    0   \n",
       "108  MGLPNPKTRKPTASEVAEWAKSNIGKRINIDNYRGSQCWDTPNYIF...    0    0    0    0   \n",
       "109  MEGNFKNVKKFIYEGEEYTKVYAGNIQVWKKPSSFVIKPLPKNKYP...    0    0    0    0   \n",
       "110  MRISQTNLGSSDLGIAGVNSGVASKNINFSNPSGMLYVTISDVYQDLRH    0    0    0    0   \n",
       "\n",
       "     AAF  AAG  ...  YYM  YYN  YYP  YYQ  YYR  YYS  YYT  YYV  YYW  YYY  \n",
       "0      0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "1      0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "2      0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "3      0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "4      0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "106    0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "107    0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "108    0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "109    0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "110    0    0  ...    0    0    0    0    0    0    0    0    0    0  \n",
       "\n",
       "[111 rows x 8004 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df.drop(['Sequence', 'Organism'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Length</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAD</th>\n",
       "      <th>AAE</th>\n",
       "      <th>AAF</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAH</th>\n",
       "      <th>AAI</th>\n",
       "      <th>...</th>\n",
       "      <th>YYM</th>\n",
       "      <th>YYN</th>\n",
       "      <th>YYP</th>\n",
       "      <th>YYQ</th>\n",
       "      <th>YYR</th>\n",
       "      <th>YYS</th>\n",
       "      <th>YYT</th>\n",
       "      <th>YYV</th>\n",
       "      <th>YYW</th>\n",
       "      <th>YYY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>284</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>holin</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>holin</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>holin</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>holin</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>629</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>629</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>endolysin</td>\n",
       "      <td>624</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>holin</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>holin</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 8002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Protein names  Length  AAA  AAC  AAD  AAE  AAF  AAG  AAH  AAI  ...  YYM  \\\n",
       "0       endolysin     284    0    0    0    0    0    0    0    0  ...    0   \n",
       "1           holin     140    0    0    0    0    0    0    0    0  ...    0   \n",
       "2           holin     140    0    0    0    0    0    0    0    0  ...    0   \n",
       "3           holin     167    0    0    0    0    0    0    0    0  ...    0   \n",
       "4           holin     167    0    0    0    0    0    0    0    0  ...    0   \n",
       "..            ...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "106     endolysin     629    0    0    0    0    0    0    0    0  ...    0   \n",
       "107     endolysin     629    0    0    0    0    0    0    0    0  ...    0   \n",
       "108     endolysin     624    0    0    0    0    0    0    0    0  ...    0   \n",
       "109         holin      57    0    0    0    0    0    0    0    0  ...    0   \n",
       "110         holin      49    0    0    0    0    0    0    0    0  ...    0   \n",
       "\n",
       "     YYN  YYP  YYQ  YYR  YYS  YYT  YYV  YYW  YYY  \n",
       "0      0    0    0    0    0    0    0    0    0  \n",
       "1      0    0    0    0    0    0    0    0    0  \n",
       "2      0    0    0    0    0    0    0    0    0  \n",
       "3      0    0    0    0    0    0    0    0    0  \n",
       "4      0    0    0    0    0    0    0    0    0  \n",
       "..   ...  ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "106    0    0    0    0    0    0    0    0    0  \n",
       "107    0    0    0    0    0    0    0    0    0  \n",
       "108    0    0    0    0    0    0    0    0    0  \n",
       "109    0    0    0    0    0    0    0    0    0  \n",
       "110    0    0    0    0    0    0    0    0    0  \n",
       "\n",
       "[111 rows x 8002 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "enc=OneHotEncoder(sparse_output=False, handle_unknown='ignore')\n",
    "df['Protein names']=enc.fit_transform(np.array(df['Protein names']).reshape(-1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 0.])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Protein names'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Length</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAD</th>\n",
       "      <th>AAE</th>\n",
       "      <th>AAF</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAH</th>\n",
       "      <th>AAI</th>\n",
       "      <th>...</th>\n",
       "      <th>YYM</th>\n",
       "      <th>YYN</th>\n",
       "      <th>YYP</th>\n",
       "      <th>YYQ</th>\n",
       "      <th>YYR</th>\n",
       "      <th>YYS</th>\n",
       "      <th>YYT</th>\n",
       "      <th>YYV</th>\n",
       "      <th>YYW</th>\n",
       "      <th>YYY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>284</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>1.0</td>\n",
       "      <td>629</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1.0</td>\n",
       "      <td>629</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>1.0</td>\n",
       "      <td>624</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>0.0</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>0.0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 8002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Protein names  Length  AAA  AAC  AAD  AAE  AAF  AAG  AAH  AAI  ...  YYM  \\\n",
       "0              1.0     284    0    0    0    0    0    0    0    0  ...    0   \n",
       "1              0.0     140    0    0    0    0    0    0    0    0  ...    0   \n",
       "2              0.0     140    0    0    0    0    0    0    0    0  ...    0   \n",
       "3              0.0     167    0    0    0    0    0    0    0    0  ...    0   \n",
       "4              0.0     167    0    0    0    0    0    0    0    0  ...    0   \n",
       "..             ...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "106            1.0     629    0    0    0    0    0    0    0    0  ...    0   \n",
       "107            1.0     629    0    0    0    0    0    0    0    0  ...    0   \n",
       "108            1.0     624    0    0    0    0    0    0    0    0  ...    0   \n",
       "109            0.0      57    0    0    0    0    0    0    0    0  ...    0   \n",
       "110            0.0      49    0    0    0    0    0    0    0    0  ...    0   \n",
       "\n",
       "     YYN  YYP  YYQ  YYR  YYS  YYT  YYV  YYW  YYY  \n",
       "0      0    0    0    0    0    0    0    0    0  \n",
       "1      0    0    0    0    0    0    0    0    0  \n",
       "2      0    0    0    0    0    0    0    0    0  \n",
       "3      0    0    0    0    0    0    0    0    0  \n",
       "4      0    0    0    0    0    0    0    0    0  \n",
       "..   ...  ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "106    0    0    0    0    0    0    0    0    0  \n",
       "107    0    0    0    0    0    0    0    0    0  \n",
       "108    0    0    0    0    0    0    0    0    0  \n",
       "109    0    0    0    0    0    0    0    0    0  \n",
       "110    0    0    0    0    0    0    0    0    0  \n",
       "\n",
       "[111 rows x 8002 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "x=df.drop('Protein names', axis=1)\n",
    "y=df['Protein names']\n",
    "X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.4, random_state=42)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(111, 8001)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "svc=SVC()\n",
    "svc.fit(X_train, y_train)\n",
    "y_pred=svc.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9333333333333333\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test,y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision: 0.93\n",
      "Recall: 1.00\n",
      "F1 Score: 0.96\n",
      "ROC AUC: 0.62\n",
      "Precision-Recall AUC: 0.97\n"
     ]
    }
   ],
   "source": [
    "precision = precision_score(y_test, y_pred)\n",
    "recall = recall_score(y_test, y_pred)\n",
    "f1 = f1_score(y_test, y_pred)\n",
    "fpr, tpr, _ = roc_curve(y_test, y_pred)\n",
    "roc_auc = auc(fpr, tpr)\n",
    "precision_vals, recall_vals, _ = precision_recall_curve(y_test, y_pred)\n",
    "pr_auc = auc(recall_vals, precision_vals)\n",
    "\n",
    "# Print other metrics\n",
    "print(f\"Precision: {precision:.2f}\")\n",
    "print(f\"Recall: {recall:.2f}\")\n",
    "print(f\"F1 Score: {f1:.2f}\")\n",
    "print(f\"ROC AUC: {roc_auc:.2f}\")\n",
    "print(f\"Precision-Recall AUC: {pr_auc:.2f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_length=len(prediction_sequence)\n",
    "prediction_sequence=\"MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTAKKEDNTPTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTGDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKSSGSLPKKKGQTSKSNIGKTFDFNGLSINVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW\"\n",
    "def prediction_feature_vector(sequence, n):\n",
    "    ngram_freq = calculate_ngram_frequencies(sequence, n)\n",
    "    vector = feature_vector.copy()\n",
    "    # Updating the feature vector with the frequency of each n-gram\n",
    "    for ngram, freq in ngram_freq.items():\n",
    "        vector[ngram] = freq\n",
    "    return pd.Series(vector)\n",
    "pred_vec=(prediction_feature_vector(prediction_sequence, 3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_vec=pd.DataFrame(np.array(pred_vec).reshape(1,-1))\n",
    "length=pd.DataFrame(np.array([seq_length]))\n",
    "pred_vec=pd.concat([pred_vec, length], axis=\"columns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 8001)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_vec.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1.]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/adminuser/.local/lib/python3.10/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "result=svc.predict(pred_vec)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "endolysin\n"
     ]
    }
   ],
   "source": [
    "if result[0]==1:\n",
    "    protein=\"endolysin\"\n",
    "else:\n",
    "    protein=\"holin\"\n",
    "print(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=proteins.loc[proteins['Protein names']==protein]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_list=a['Sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_list=sequences_list.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def similarity_search(sequences_list, target_sequence, top_n=5):\n",
    "    \"\"\"\n",
    "    Perform similarity search on an array of protein sequences.\n",
    "    \n",
    "    Parameters:\n",
    "    - sequences_list: list of strings, each string representing a protein sequence\n",
    "    - target_sequence: string, the protein sequence to search for\n",
    "    - top_n: int, the number of top similar sequences to return (default is 5)\n",
    "    \n",
    "    Returns:\n",
    "    - List of tuples with the most similar sequences and their similarity percentages\n",
    "    \"\"\"\n",
    "    def similarity(seq1, seq2):\n",
    "        \"\"\"Calculate similarity score using SequenceMatcher.\"\"\"\n",
    "        return SequenceMatcher(None, seq1, seq2).ratio()\n",
    "\n",
    "    # Calculate similarity for each sequence in the list\n",
    "    similarities = [(seq, similarity(target_sequence, seq) * 100) for seq in sequences_list]\n",
    "    \n",
    "    # Sort the list based on similarity scores in descending order\n",
    "    similarities.sort(key=lambda x: x[1], reverse=True)\n",
    "    \n",
    "    # Return the top N most similar sequences with their percentage match\n",
    "    return similarities[:top_n]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTAKKEDNTPTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTGDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKSSGSLPKKKGKKQTSKSNIGKTFDFNGLSINVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW, Similarity: 99.79%\n",
      "Sequence: MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTAKKEDNTPTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTGDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKSSGSLPNKKGKKQTSKSNIGKTFDFNGLSTNVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW, Similarity: 95.24%\n",
      "Sequence: MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTVKKEDNTLTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTGDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKSSGSLPKKKGKKQTSKSNIGKTFDFNGLSTNVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW, Similarity: 94.41%\n",
      "Sequence: MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTAKKEDNTPTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTGDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKNSGSLPKKKGKKQTSKSNIGKTFDFNGLSTNVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW, Similarity: 94.00%\n",
      "Sequence: MKTKKQALKWILNTIGQGIDWDKMYGFQCMDLVVAYLYYVTDGKIAMWGNAIDAPKNNFKGTAKVIKNYPAFRPEEGDIVVWSYGNFSTYGHIAVVIDGDPYGDLQYITVAEQNWNGLGLYKQEVTTKRIHNYDGVSHFIRPKFKKTVKKEDNTLTKEKNNKKTKGKKLKVSTQRINYTMDKRGYKPKFVVIHNDAGSSSAQQYEQGLKNAGYSRYAQGVAHAYASDGYVWEAISEDRIAWHTVLRTLLNISDGTNPGTGNFEGYGIEVCQSLGDRNTFLKNEQTVFQFIAEKLQKWNLPANRNTIRLHNEFIQTECPHASAYYHAGMNTKVDAYTKERQLKIKDYFIKQIRAYMKGSTPKSTVVKSSKSSGSLPKKKGKKQTSKSNIGKTFDFNGLSTNVWGTKWYYENNTFTCNARQGIITRVGSPFTTAPQAGVLFYGQTVTYNQVAVNPKEPFVWISWITNNGTEVWMPIEVLDSNNKIIEQWGTFGW, Similarity: 75.36%\n"
     ]
    }
   ],
   "source": [
    "top_similar=similarity_search(sequences_list, prediction_sequence, 5)\n",
    "for seq, score in top_similar:\n",
    "    print(f\"Sequence: {seq}, Similarity: {score:.2f}%\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
